stan/math/prim/fun/trace_gen_quad_form.hpp

```diff
@@ -34,11 +28,13 @@ namespace math {
  * @throw std::domain_error if A cannot be multiplied by B or B cannot
  * be multiplied by D.
  */
-template <typename TD, int RD, int CD, typename TA, int RA, int CA, typename TB,
-          int RB, int CB, typename = require_all_not_var_t<TD, TA, TB>>
-inline return_type_t<TD, TA, TB> trace_gen_quad_form(
-    const Eigen::Matrix<TD, RD, CD> &D, const Eigen::Matrix<TA, RA, CA> &A,
-    const Eigen::Matrix<TB, RB, CB> &B) {
+template <typename TD, typename TA, typename TB,
+          typename = require_all_not_var_t<value_type_t<TD>, value_type_t<TA>,
+                                           value_type_t<TB>>,
+          typename = require_any_not_same_t<double, value_type_t<TD>,
+                                            value_type_t<TA>, value_type_t<TB>>,
+          typename = require_all_eigen_t<TD, TA, TB>>
```

Comment:
Can we flip these? So users would read this is for 

- Eigen expressions
- Where not all value types vars
- and at least one is not a double